pandas/core/frame.py

```diff
@@ -1436,7 +1460,21 @@ def to_records(self, index=True, convert_datetime64=None):
             arrays = [self[c].get_values() for c in self.columns]
             names = lmap(compat.text_type, self.columns)
 
-        formats = [v.dtype for v in arrays]
+        if strings_as_fixed_length:
+            def is_string(arr):
```

Comment:
don't do this,  rather use

``lib.infer_dtype(arr) == 'string'`` (there is also a unicode variant, so I guess use 'U' for that)
for that matter if you actually have bytes it will return that as well.

Anything else would trigger 'O', note that this routine will correctly infer all dtypes

Reply:
e.g.

```
# this is just the public interface
In [3]: from pandas.api.types import infer_dtype

NameError: name 'ifner_dtype' is not defined

In [5]: infer_dtype(['foo', 'bar'])
Out[5]: 'string'

In [6]: infer_dtype([b'foo', b'bar'])
Out[6]: 'bytes'

# in PY2 this would be unicode
In [7]: infer_dtype([u'foo', u'bar'])
Out[7]: 'string'

In [8]: infer_dtype([u'foo', b'bar'])
Out[8]: 'mixed'
```

Reply:
@jreback Sorry for the late reply. I think the original issue #18146 was meant for ASCII-encoded strings, and your comment has made me realize that this wouldn't work for strings with unicode characters. (Numpy throws an error like `UnicodeEncodeError: 'ascii' codec can't encode character '\u212b' in position 0: ordinal not in range(128)`).

I propose that this feature be implemented for `bytes` objects instead, which are still encoded as `O` dtype but can be more efficiently encoded as `S` dtype. And the option would be called `bytes_as_fixed_length`. What do you think?